packages/muelu/test/viz/AggExport.cpp

```diff
@@ -0,0 +1,401 @@
+      }
+
+      if (!nullFile.empty())
+        nullspace = Xpetra::IO<SC,LO,GO,Node>::ReadMultiVector(nullFile, map);
+    }
+
+    comm->barrier();
+    tm = Teuchos::null;
+
+    galeriStream << "Galeri complete.\n========================================================" << std::endl;
+
+    std::string matrixType = galeriParameters.GetMatrixType();
+    std::string aggVizType = paramList.get<std::string>("aggregation: output file: agg style");
+    aggVizType.erase(std::remove_if(aggVizType.begin(),aggVizType.end(), ::isspace), aggVizType.end());
+    if(ndims == 2)
+      paramList.set<std::string>("aggregation: output filename", "MPI-Viz-Output-2D-Level%LEVELID-Proc%PROCID");
+    else if(ndims == 3){
+      if(comm->getSize() > 1)
+        paramList.set<std::string>("aggregation: output filename", "MPI-Viz-Output-"+matrixType+"-"+aggVizType+"-Level%LEVELID-Proc%PROCID");
+      else
+        paramList.set<std::string>("aggregation: output filename","MPI-Viz-Output-"+matrixType+"-"+aggVizType+"-Level%LEVELID");
+    }
+
+    if (nullspace.is_null()) {
+      int blkSize = 1;
+      if (paramList.isSublist("Matrix")) {
+        // Factory style parameter list
+        const Teuchos::ParameterList& operatorList = paramList.sublist("Matrix");
+        if (operatorList.isParameter("PDE equations"))
+          blkSize = operatorList.get<int>("PDE equations");
+
+      } else if (paramList.isParameter("number of equations")) {
+        // Easy style parameter list
+        blkSize = paramList.get<int>("number of equations");
+      }
+
+      nullspace = MultiVectorFactory::Build(map, blkSize);
+      for (int i = 0; i < blkSize; i++) {
+        RCP<const Map> domainMap = A->getDomainMap();
+        GO             indexBase = domainMap->getIndexBase();
+
+        ArrayRCP<SC> nsData = nullspace->getDataNonConst(i);
+        for (int j = 0; j < nsData.size(); j++) {
+          GO GID = domainMap->getGlobalElement(j) - indexBase;
+
+          if ((GID-i) % blkSize == 0)
+            nsData[j] = Teuchos::ScalarTraits<SC>::one();
+        }
+      }
+    }
+
+    // Instead of checking each time for rank, create a rank 0 stream
+    RCP<Teuchos::FancyOStream> fancy = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+    Teuchos::FancyOStream& out = *fancy;
+    out.setOutputToRootOnly(0);
+
+    out << galeriStream.str();
+
+    // =========================================================================
+    // Preconditioner construction
+    // =========================================================================
+    comm->barrier();
+    tm.reset();
+    tm = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("ScalingTest: 1.5 - MueLu read XML")));
+
+    RCP<HierarchyManager> mueLuFactory = rcp(new ParameterListInterpreter(paramList));
+
+    comm->barrier();
+    tm.reset();
+    tm = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("ScalingTest: 2 - MueLu Setup")));
+
+    RCP<Hierarchy> H;
+    for (int i = 0; i <= numRebuilds; i++) {
```

Comment:
What's the purpose of rebuilding the hierarchy here? If there's none, this loop can be removed. If so, please also remove the command line argument to control the number of rebuilds.